jenkins/ci.suse.de/pipelines/openstack-ardana-vcloud.Jenkinsfile

```diff
@@ -0,0 +1,214 @@
+  }
+
+  stages {
+    stage('setup workspace and environment') {
+      steps {
+        cleanWs()
+
+        // If the job is set up to reuse an existing workspace, replace the
+        // current workspace with a symlink to the reused one.
+        // NOTE: even if we specify the reused workspace as the
+        // customWorkspace variable value, Jenkins will refuse to reuse a
+        // workspace that's already in use by one of the currently running
+        // jobs and will just create a new one.
+        sh '''
+          if [ -n "${reuse_workspace}" ]; then
+            rmdir "${WORKSPACE}"
+            ln -s "${reuse_workspace}" "${WORKSPACE}"
+          fi
+        '''
+
+        script {
+          if ("${label}" != '') {
+            currentBuild.displayName = "#${BUILD_NUMBER} ${label}"
+            //currentBuild.description = ""
+            env.heat_stack_name="$JOB_NAME-$label"
+          }
+          else {
+            currentBuild.displayName = "${JOB_NAME}-${BUILD_NUMBER}"
+            //currentBuild.description = ""
+            env.heat_stack_name="$JOB_NAME-$BUILD_NUMBER"
+          }
+          env.input_model_path = "${WORKSPACE}/input-model"
+          env.heat_template_file = "${WORKSPACE}/heat-ardana-${model}.yaml"
+        }
+      }
+    }
+
+    stage('clone automation repo') {
+      when {
+        expression { reuse_workspace == '' }
+      }
+      steps {
+        dir("automation-git") {
+          checkout scm
+        }
+      }
+    }
+
+    stage('clone input model repo') {
+      when {
+        expression { git_input_model_repo != '' && scenario == '' }
+      }
+      steps {
+        dir('input-model-git') {
+          git(
+             url: "${git_input_model_repo}",
+             branch: "${git_input_model_branch}"
+          )
+        }
+        script {
+          env.input_model_path = "${WORKSPACE}/input-model-git/${git_input_model_path}/${model}"
+        }
+      }
+    }
+
+    stage('generate input model') {
+      when {
+          expression { scenario != '' }
+      }
+      steps {
+        dir('automation-git/scripts/jenkins/ardana/ansible') {
+          script {
+            env.virt_config = "${WORKSPACE}/${scenario}-virt-config.yml"
+          }
+          sh '''
+            source /opt/ansible/bin/activate
+            ansible-playbook -v \
+                             -e cloud_release="${cloud_release}" \
+                             -e scenario_name="${scenario}" \
+                             -e input_model_dir="${input_model_path}" \
+                             -e virt_config_file="${virt_config}" \
+                             -e clm_model=$clm_model \
+                             -e controllers=$controllers \
+                             -e sles_computes=$sles_computes \
+                             -e rhel_computes=$rhel_computes \
+                             generate-input-model.yml
+          '''
+        }
+      }
+    }
+
+    stage('generate heat') {
```

Comment:
This is just a random location in the pull changes to group this discussion.
### looser coupling with Jenkins
I'd like for most of what is done in Jenkins jobs to work outside of a Jenkins server (say on a development machine). For normal Jobs that is a matter of just using one JJB builder that calls a script file instead of inlining scripts in the JJB job definitions. https://github.com/SUSE-Cloud/automation/pull/2671/ makes a change in that direction.

For Jenkins pipelines this is not as easy as Jenkins needs to know at least the steps, their order and which ones can be run in parallel. Duplicating this information in a script file that can be executed and a Jenkinsfile has the downside of duplication, which I'd like to avoid.

Maybe there is already a way to run a Jenkins pipeline without a Jenkins server installation?

### Idea how to solve this:
Lets say we have cloud.py (could be acloud, mkcloud, an ansible playbook/extension, a yml file with some python code to run it) which does what the Jenkins job should do.

We implement cloud.py --generate-jenkins which internally uses the same information it would when running by itself to generate the Jenkinsfile. This might be internally represented as a list of steps with a special parallel step that contains a list of steps to execute in parallel. This internal representation can then be used to run them all. And to generate a CLI to call them individually like testcloud.py --step step1. The CLI for an individual step is then used in the generated Jenkins file to run the code for the individual steps.

To call the Jenkins pipeline we have a normal job that checks out the repo where testcloud.py exists, then generates the Jenkinsfile with it, then create the Jenkins subjob to run the pipeline file that was just generated.

Reply:
Just posting this here so I'll find this again. Example what it might look like in Ansible:
```
- import_playbook: "{{ item.playbook }}"
  when: "{{ item.when }}"
  loop:
  - { playbook: generate-input-model.yml, when: True }
  - { playbook: rebuild-deployer-vm.yml, when: True }
  - { playbook: bootstrap-deployer-vm.yml, when: True }
  - { playbook: ardana-ses-integration.yml, when: ses_enabled }
  - { playbook: shutdown-bm-nodes.yml, when: True }
  - { playbook: run-ardana-deploy-plays.yml, when: True }
  - { playbook: ardana-ses-rgw.yml, when: ses_enabled }
```
Though `include_playbook` might be needed. Unclear if this will work even then. Also it was said that `when` might not work well. Roles are probably better than using playbooks.